### ISSUE TITLE
Warn for overlays with different dimensions?

### ISSUE THREAD

**Issue by jbednar** [ioam/holoviews#1751]
Right now, overlays seem to use the name of the first object's dimensions, but happily combine dimensions with different names:

![image](https://user-images.githubusercontent.com/1695496/28590159-c11e2c58-7146-11e7-95a3-5f1dc8b98e68.png)

This behavior is a bit surprising to me; I'd have expected a warning or else the axis label to list all the dimensions that apply to this plot (i.e. "horiz, h" here).  Not sure how widespread such overlaying is, i.e. how many existing usages adding such a warning would affect, but it seems like people must already have to deal with this because whether the displayed axis label is the one they wanted seems like only a matter of chance.

**Comment by jlstevens** [comment]
I'm open to a issuing a warning when overlaying non-matching dimensions. @philippjfr?

**Comment by philippjfr** [comment]
>I'm open to a issuing a warning when overlaying non-matching dimensions. @philippjfr?

On the one hand this would force people to give their objects proper ``kdims`` on the other hand I think this happens all the time and could in some cases be quite annoying. In theory we could also add parallel axes eventually. I'll have to think about it and maybe try the warning out for a bit.

**Comment by jlstevens** [comment]
The warning can be a config option so we can try it and then decided whether it should be on or off by default.

**Comment by jbednar** [comment]
If it happens all the time, it means that people's plots are being labeled with some arbitrary dimension, and they probably just don't notice that it's only the name of one of the things they are plotting, not all of them.  So I'd vote for either having a warning, or showing all the dimensions in a list, so that it's not silently being misleading.

**Comment by philippjfr** [comment]
I almost never give annotations or annotation-like things proper ``kdims``. My compromise would be to warn for non-default dimensions.

**Comment by jbednar** [comment]
I agree, for annotations; I'm talking about actual overlaid data.  Sure, if there's more than one non-default dimension (as above), then it's a clear case for warning.

**Comment by johnzzzz** [comment]
An Overlay's x and y axes have Dimensions that seem be created ONLY through inference on it contained Elements. In jbednars case, the inference picks the first contained Element and does not detect conflicts between other elements. I think a warning would be useful in this case.

But also, there does not seem to be a way to explicitly set the Dimensions of Overlay axes. Allowing the user to override how the inferences implemented, would be a more general solution.

Making inferences a first class object would be even a better solution.

**Comment by jbednar** [comment]
I still don't get it; how is the behavior of arbitrarily choosing "horiz" in the above case, instead of "h", more convenient?  To me it is simply surprising and confusing, and hard to remember. Even if this were to be a general rule, why should the label be taken from the *first* item in the overlay, given that Overlays generally leave the *last* thing on top of the plot?  I think the current behavior is just a bug, for anything with specified (non-default) dimensions; I really can't see any justification for it.

**Comment by jlstevens** [comment]
I think I would be ok with issuing warnings as long as you can disable them to allow the old behavior without the warnings. @philippjfr ?

**Comment by philippjfr** [comment]
Warnings for non-default dimensions are probably okay, but I think it would have to be coupled with detailed documentation about working with wide datasets and how to make your data tidy. I'd also be okay with using the name of the first layer with a non-default dimension for the axis.

**Comment by jlstevens** [comment]
>  I'd also be okay with using the name of the first layer with a non-default dimension for the axis.

I would also be happy with that policy.
